### PR TITLE
Fix unhandled exceptions in MockAnnotationProcessor

### LINKED ISSUE #38880: [Bug]: Unhandled exceptions in MockAnnotationProcessor related to Function Mocking
### Description

There are a several unhandled exceptions in the class MockAnnotationProcessor that needs to be fixed.
1. For single file projects, function mocking is not supported. Should provide a proper diagnostic for that. This is also tracked by https://github.com/ballerina-platform/ballerina-lang/issues/38720
2. When the provided `functionName` for function mocking is not found in the project, a proper diagnostic should be given. A part of this is fixed with https://github.com/ballerina-platform/ballerina-lang/issues/38693. However there are cases where the fix is not applicable to.
    ```ballerina
    @test:Mock { functionName: "intAdd1" }  // intAdd1 is non-existent
    test:MockFunction intAddMockFn = new();
    ```
  3. A function to be mocked should be in a file that is not a test file. If we do not find a function with the name provided with `functionName` in the non-test files we do not give a proper error diagnostic as long as a function with that name is in a test file. However this leads to the following runtime error.
    ``` bash
        [2022-11-30 09:51:02,866] SEVERE {b7a.log.crash} - Mock Function and Function to Mock cannot be null 
        java.lang.IllegalStateException: Mock Function and Function to Mock cannot be null
                at org.wso2.ballerinalang.compiler.desugar.MockDesugar.generateMockFunction(MockDesugar.java:162)
                at org.wso2.ballerinalang.compiler.desugar.MockDesugar.generateMockFunctions(MockDesugar.java:117)
    ```

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Test Framework

### Related issue(s) (optional)

https://github.com/ballerina-platform/ballerina-lang/issues/38720
https://github.com/ballerina-platform/ballerina-lang/issues/38693

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by gayaldassanayake**
## Purpose

$ subject

Fixes #38880 , https://github.com/ballerina-platform/ballerina-lang/issues/38720

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/core/MockAnnotationProcessor.java [hunk 1]
```diff
@@ -320,33 +331,34 @@ private String formatPackageName(String value, BLangPackage parent) {
      *
      * @param functionName   Name of the function to mock
      * @param attachmentNode MockFunction object attachment node
+     * @return true if the provided function name valid
      */
-    private void validateFunctionName(String functionName, PackageID functionToMockID,
+    private boolean validateFunctionName(String functionName, String moduleName, PackageID functionToMockID,
```
Comment (Dilhasha) [inline misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/core/MockAnnotationProcessor.java:336]:
```suggestion
    private boolean isValidFunctionName(String functionName, String moduleName, PackageID functionToMockID,
```

File: misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/core/MockAnnotationProcessor.java [hunk 2]
```diff
@@ -207,7 +220,7 @@ public void process(FunctionNode functionNode, List<AnnotationAttachmentNode> an
                     PackageID functionToMockID = getPackageID(vals[0]);
                     if (functionToMockID == null) {
                         diagnosticLog.logDiagnostic(DiagnosticSeverity.ERROR, attachmentNode.getPosition(),
-                                "could not find module specified");
+                                "could not find module specified '" + vals[0] + "'");
```
Comment (Dilhasha) [inline misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/core/MockAnnotationProcessor.java:223]:
```suggestion
                                "could not find specified module '" + vals[0] + "'");
```

File: misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/core/MockAnnotationProcessor.java [hunk 3]
```diff
@@ -218,13 +231,13 @@ public void process(FunctionNode functionNode, List<AnnotationAttachmentNode> an
...
-                                    "incompatible types: expected " + functionToMockType.toString()
-                                            + " but found " + mockFunctionType.toString());
+                                    "incompatible types: expected " + functionToMockType
+                                            + " but found " + mockFunctionType);
                             break;
                         }
                     } else {
                         diagnosticLog.logDiagnostic(DiagnosticSeverity.ERROR, attachmentNode.getPosition(),
-                                "could not find functions in module");
+                                "could not find function " + vals[1] + " in module '" + vals[0] + "'");
```
Comment (Dilhasha) [inline misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/core/MockAnnotationProcessor.java:240]:
```suggestion
                                "could not find function '" + vals[1] + "' in module '" + vals[0] + "'");
```

File: misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/core/MockAnnotationProcessor.java [hunk 4]
```diff
@@ -109,6 +111,13 @@ public void process(SimpleVariableNode simpleVariableNode, List<AnnotationAttach
         annotations = annotations.stream().distinct().collect(Collectors.toList());
         // Iterate through all the annotations
         for (AnnotationAttachmentNode attachmentNode : annotations) {
+            // Check if the package belongs to a single file project
+            if (packageName.equals(DOT)) {
+                diagnosticLog.logDiagnostic(
+                        DiagnosticSeverity.ERROR, (ModuleDescriptor) null, attachmentNode.getPosition(),
+                        "Function mocking is not supported for single file projects");
```
Comment (Thevakumar-Luheerathan) [inline misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/core/MockAnnotationProcessor.java:118]: As far as I know we need start in lower case for an error. Please verify that.

**Comment by azinneera** [comment]
Shall we address https://discord.com/channels/957996897782616114/1017493204326686860/threads/1049569659466297414 also in the same PR?

**Review by azinneera — APPROVED** [review]

**Pull request merged**